api/controllers/FlairController.js

```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
 
 module.exports = {
 
-  apply: function (req, res) {
-    var appData = {
-      user: req.user.name,
-      flair: req.allParams().flair,
-      sub: req.allParams().sub
-    };
-
-    Application.find(appData).exec(function (err, app) {
-      if (err) {
-        return res.serverError(err);
+  apply: async function (req, res) {
+    try {
+      var allFlairs = await Flair.find();
+      var userRefs = await Reference.find({user: req.user.name});
+      var appData = {user: req.user.name, flair: req.allParams().flair, sub: req.allParams().sub};
+      if (await Application.findOne(appData)) {
+        return res.status(400).json({error: 'You have already applied for that flair'});
       }
-      if (app.length > 0) {
-        return res.badRequest("Application already exists");
+      var flairIndex = allFlairs.map(function (flairObj) {
+        return flairObj.name;
+      }).indexOf(req.allParams().flair);
+      if (flairIndex === -1) {
+        return res.status(400).json({error: 'That flair does not exist'});
       }
-      Application.create(appData).exec(function (err, apps) {
-        if (err) {
-          return res.serverError(err);
-        }
-        if (apps) {
-          return res.ok(appData);
-        }
-      });
-    });
+      var applicationFlair = allFlairs[flairIndex];
+      if (!Flairs.canUserApply(userRefs, applicationFlair, Flairs.getUserFlairs(req.user, allFlairs))) {
```

Comment:
We could probably add a wrapper function inside Flairs service that we can pass the name into rather than having to do the grabbing of the actual flair out here, but it's not important.